components/ingest-service/pipeline/publisher/msg_distributor.go

```diff
@@ -0,0 +1,87 @@
+package publisher
+
+import (
+	"time"
+
+	"github.com/chef/automate/components/ingest-service/pipeline/message"
+)
+
+// BuildMsgDistributor a message distributor that sends all messages to the first child pipe until it is
+// full then start sending messages to the next child pipe in the line. When all the pipes are full the thread waits and
+// then checks to pipes again until it can place the message in a child pipe.
+func BuildMsgDistributor(childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) message.ChefRunPipe {
+	return func(in <-chan message.ChefRun) <-chan message.ChefRun {
+		return msgDistributor(in, childPipeBuilder, numProcessors, childPipeInboxSize)
+	}
+}
+
+func msgDistributor(in <-chan message.ChefRun,
+	childPipeBuilder message.ChefRunPipe, numProcessors, childPipeInboxSize int) <-chan message.ChefRun {
+	pipeInChannels, out := buildChildPipes(childPipeBuilder, numProcessors, childPipeInboxSize)
+
+	go func() {
+		for msg := range in {
+			sendMessage(pipeInChannels, msg)
+		}
+		close(out)
+	}()
+
+	return out
+}
+
+func sendMessage(pipeInChannels []chan message.ChefRun, msg message.ChefRun) {
+	for true {
```

Comment:
[nit]
```suggestion
	for {
```

Reply:
Not sure how we feel about this style, but I think if you wanted you could do:

```golang
for !distributeMessage(pipeInChannels, msg) {
	// All pipes are full. Wait and try again
	time.Sleep(time.Millisecond * 10)
}

Reply:
Yeah, I thought about doing that change, but it does not seem as readable. The "distributeMessage" function does not say when to stop the for-loop. 
What about?
```suggestion
for messageProcessed := distributeMessage(pipeInChannels, msg); !messageProcessed; messageProcessed = distributeMessage(pipeInChannels, msg) {
   // All pipes are full. Wait and try again
   time.Sleep(time.Millisecond * 10)
}
```

Reply:
Eh, I'd just leave it, it was pretty clear as is, I was just musing aloud :D